pandas/core/dtypes/concat.py

```diff
@@ -175,7 +175,7 @@ def is_nonempty(x):
         return _concat_sparse(to_concat, axis=axis, typs=typs)
 
     extensions = [is_extension_array_dtype(x) for x in to_concat]
-    if any(extensions):
+    if any(extensions) and axis == 1:
         to_concat = [np.atleast_2d(x.astype('object')) for x in to_concat]
```

Comment:
I find this very strange that you need to add axis=1 here, I don't think this routine should be call at all if axis=1. This seems like it should be handled at a higher level

Reply:
>  I find this very strange that you need to add axis=1 here, I don't think this routine should be call at all if axis=1

`axis=1` is used all over `_concat_compat`.  Why do you think it's strange?

For the specific case from #20832 it is `axis=0`, but we don't want to reshape to 2d.

Reply:
Yes, the special casing non-consolidateble blocks with `axis=1` is indeed an existing pattern. 

For categoricals:

https://github.com/pandas-dev/pandas/blob/563a6ad108a79b0192a25d51d743812a18e21b15/pandas/core/dtypes/concat.py#L220-L221

for datetimetz:

https://github.com/pandas-dev/pandas/blob/563a6ad108a79b0192a25d51d743812a18e21b15/pandas/core/dtypes/concat.py#L453-L454

Note that this is all rather confusing, as `axis=1` in the "concatting code" actually means `axis=0` in user facing code (because we store the 1D data in 2D objects):

https://github.com/pandas-dev/pandas/blob/563a6ad108a79b0192a25d51d743812a18e21b15/pandas/core/reshape/concat.py#L312-L315

Reply:
right this should be changed to use ``_get_block_manager_axis`` FYI (but can be in future).

Reply:
> right this should be changed to use _get_block_manager_axis FYI (but can be in future).

I don't think so, `_get_block_manager_axis` is for a full Series/DataFrame, the logic here is dtype-dependent

Reply:
what does your respsone have to do with my comment? this is exactly what _get_block_managed_axis does

Reply:
OK, my response was not very well phrased, but still, then you will have to explain your comment better :-) 

I understood your comment as: "we should use `_get_block_manager_axis` here ", and AFAIK, `_get_block_manager_axis` lets you convert an axis of 0/1 to the appropriate number for blocks (so for DataFrame it switches 0 and 1). But here, I don't convert any `axis` argument, I reshape data depending on the `axis` value (and depending on the dtype). 
So I don't fully understand where in the above code lines I would use ``_get_block_manager_axis``

Reply:
```
diff --git a/pandas/core/reshape/concat.py b/pandas/core/reshape/concat.py
index 6e564975f..6570e7a83 100644
--- a/pandas/core/reshape/concat.py
+++ b/pandas/core/reshape/concat.py
@@ -312,7 +312,7 @@ class _Concatenator(object):
         # Need to flip BlockManager axis in the DataFrame special case
         self._is_frame = isinstance(sample, DataFrame)
         if self._is_frame:
-            axis = 1 if axis == 0 else 0
+            axis = sample._get_block_manager_axis(axis)
 
         self._is_series = isinstance(sample, Series)
         if not 0 <= axis <= sample.ndim:
```

is a more correct usage, though prob doesn't simplify code much. this is spagetti anyhow.

Reply:
> this is spagetti anyhow.

On that we certainly agree! :-)